adding semi-supernode feature

james-prysm opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, check out our contribution guide here https://docs.prylabs.network/docs/contribute/contribution-guidelines
   You will then need to sign our Contributor License Agreement (CLA), which will show up as a comment from a bot in this pull request after you open it. We cannot review code without a signed CLA.
2. Please file an associated tracking issue if this pull request is non-trivial and requires context for our team to understand. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
5. A changelog entry is required for user facing issues.
-->

**What type of PR is this?**

Feature


**What does this PR do? Why is it needed?**

| Feature                 | Semi-Supernode            | Supernode                |
| ----------------------- | ------------------------- | ------------------------ |
| **Custody Groups**      | 64                        | 128                      |
| **Data Columns**        | 64                        | 128                      |
| **Storage**             | ~50%                      | ~100%                    |
| **Blob Reconstruction** | Yes (via Reed-Solomon)    | No reconstruction needed |
| **Flag**                | `--semi-supernode`        | `--supernode`            |
| **Can serve all blobs** | Yes (with reconstruction) | Yes (directly)           |

**note** if your validator total effective balance results in more custody than the semi-supernode it will override those those requirements.

cgc=64 from @nalepae 
Pro:
- We are useful to the network
- Less disconnection likelihood
- Straight forward to implement

Con:
- We cannot revert to a full node
- We have to serve incoming RPC requests corresponding to 64 columns

Tested the following using this kurtosis setup

```
participants:
  # Super-nodes
  - el_type: geth
    el_image: ethpandaops/geth:master
    cl_type: prysm
    vc_image: gcr.io/offchainlabs/prysm/validator:latest
    cl_image: gcr.io/offchainlabs/prysm/beacon-chain:latest
    count: 2
    cl_extra_params:
      - --supernode
    vc_extra_params:
      - --verbosity=debug
  # Full-nodes
  - el_type: geth
    el_image: ethpandaops/geth:master
    cl_type: prysm
    vc_image: gcr.io/offchainlabs/prysm/validator:latest
    cl_image: gcr.io/offchainlabs/prysm/beacon-chain:latest
    count: 2
    validator_count: 1
    cl_extra_params:
      - --semi-supernode
    vc_extra_params:
      - --verbosity=debug

additional_services:
  - dora
  - spamoor

spamoor_params:
  image: ethpandaops/spamoor:master
  max_mem: 4000
  spammers:
    - scenario: eoatx
      config:
        throughput: 200
    - scenario: blobs
      config:
        throughput: 20

network_params:
  fulu_fork_epoch: 0
  withdrawal_type: "0x02"
  preset: mainnet

global_log_level: debug
```

```
curl -H "Accept: application/json" http://127.0.0.1:32961/eth/v1/node/identity
{"data":{"peer_id":"16Uiu2HAm7xzhnGwea8gkcxRSC6fzUkvryP6d9HdWNkoeTkj6RSqw","enr":"enr:-Ni4QIH5u2NQz17_pTe9DcCfUyG8TidDJJjIeBpJRRm4ACQzGBpCJdyUP9eGZzwwZ2HS1TnB9ACxFMQ5LP5njnMDLm-GAZqZEXjih2F0dG5ldHOIAAAAAAAwAACDY2djQIRldGgykLZy_whwAAA4__________-CaWSCdjSCaXCErBAAE4NuZmSEAAAAAIRxdWljgjLIiXNlY3AyNTZrMaECulJrXpSOBmCsQWcGYzQsst7r3-Owlc9iZbEcJTDkB6qIc3luY25ldHMFg3RjcIIyyIN1ZHCCLuA","p2p_addresses":["/ip4/172.16.0.19/tcp/13000/p2p/16Uiu2HAm7xzhnGwea8gkcxRSC6fzUkvryP6d9HdWNkoeTkj6RSqw","/ip4/172.16.0.19/udp/13000/quic-v1/p2p/16Uiu2HAm7xzhnGwea8gkcxRSC6fzUkvryP6d9HdWNkoeTkj6RSqw"],"discovery_addresses":["/ip4/172.16.0.19/udp/12000/p2p/16Uiu2HAm7xzhnGwea8gkcxRSC6fzUkvryP6d9HdWNkoeTkj6RSqw"],"metadata":{"seq_number":"3","attnets":"0x0000000000300000","syncnets":"0x05","custody_group_count":"64"}}}
```

```
curl -s http://127.0.0.1:32961/eth/v1/debug/beacon/data_column_sidecars/head | jq '.data | length'
64
```

```
curl -X 'GET' \
  'http://127.0.0.1:32961/eth/v1/beacon/blobs/head' \
  -H 'accept: application/json'
  ```

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

james-prysm committed:
adding semi-supernode feature

james-prysm committed:
gaz

james-prysm committed:
changelog

james-prysm committed:
Merge branch 'develop' into semi-supernode

james-prysm committed:
handle case if validators requirements surpass semi node requirements

james-prysm committed:
handle case if validators requirements surpass semi node requirements

james-prysm committed:
making more idiomatic

james-prysm committed:
fixing code coverage

james-prysm committed:
forgot to register flag

james-prysm committed:
making sure semi-supernode can't be downgraded

james-prysm commented on beacon-chain/sync/subscriber.go:
```diff
@@ -712,7 +713,28 @@ func (s *Service) samplingSize() (uint64, error) {
+			}).Warn("Validator custody requirement exceeds semi-supernode minimum. Subscribing to additional subnets.")
+			return requiredSampling, nil
+		}
+		return semiSupernodeMin, nil
```
Wondering if this one is required or if the p2p custody count part already handles. added due to 
```
if flags.Get().SubscribeAllDataSubnets {
		return cfg.DataColumnSidecarSubnetCount, nil
	}
	``

james-prysm replied:
somehow passed other tests without this change

james-prysm committed:
linting

james-prysm committed:
making code more idiomatic

james-prysm committed:
linting

james-prysm committed:
fixing test

james-prysm committed:
Merge branch 'develop' into semi-supernode

james-prysm committed:
gofmt

james-prysm committed:
simplifying samplingSize function and fixing bug in service

james-prysm commented on beacon-chain/sync/subscriber.go:
```diff
@@ -692,26 +692,28 @@ func (s *Service) dataColumnSubnetIndices(primitives.Slot) map[uint64]bool {
 func (s *Service) samplingSize() (uint64, error) {
 	cfg := params.BeaconConfig()
 
-	if flags.Get().SubscribeAllDataSubnets {
```
This is a redundant check that makes it a bit more confusing

james-prysm replied:
already handled in p2p.CustodyGroupCounts

james-prysm committed:
Merge branch 'develop' into semi-supernode

james-prysm commented:
Tested with and without the flag. adding and removing validators on hoodi, have not tested manually but in unit test if validator requirements go over the semi-supernode limit that it will override the semi-supernode mode

prestonvanloon commented on cmd/beacon-chain/flags/config.go:
```diff
@@ -56,6 +57,11 @@ func ConfigureGlobalFlags(ctx *cli.Context) error {
 	}
 
+	if ctx.Bool(SemiSupernode.Name) {
+		log.Warning("Operating in semi-supernode mode (64 data columns)")
```
Not sure you want to mention explicitly `64` here since it could change?


```suggestion
		log.Warning("Operating in semi-supernode mode (retain 50% of data columns)")
```

prestonvanloon commented on beacon-chain/blockchain/service.go:
```diff
@@ -490,10 +491,42 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
+		)
+		// Note: Semi-supernode mode is persisted in the database and will be enforced
+		// by the subscriber service when determining samplingSize()
+	}
```
Why can't the user go back to a normal full node?

james-prysm replied:
mostly from @nalepae 's comment
> cgc=64
Pro:
We are useful to the network
Less disconnection likelihood
Straight forward to implement
Con:
We cannot revert to a full node
We have to serve incoming RPC requests corresponding to 64 columns

however lighthouse does not allow reversion either

nalepae replied:
It's actually [the specification](https://github.com/ethereum/consensus-specs/blob/master/specs/fulu/validator.md#validator-custody):

> If a node's custody requirements decrease, it SHOULD NOT update the custody_group_count to reflect this reduction. The node SHOULD continue to custody and advertise the previous (highest) custody_group_count. The node SHOULD continue to respond to any DataColumnSidecar request corresponding to the previous (highest) custody_group_count. The previous (highest) custody_group_count SHOULD persist across node restarts.

It's one of the main con of the `cgc=64` solution compared to `cgc=4 and sampling=60` solution, where going back to a full node is possible.

james-prysm replied:
I've removed db update for semi supernode while keeping it for super node to simplify for this pr https://github.com/OffchainLabs/prysm/pull/16029/commits/80e778044791736be2ef8b2e1aed549a6cf7e81c

prestonvanloon commented on beacon-chain/sync/custody.go:
```diff
@@ -103,12 +103,35 @@ func (s *Service) custodyGroupCount(context.Context) (uint64, error) {
+		return effectiveCustodyRequirement, nil
+	}
+
+	// Semi-supernode mode custodies exactly half of all custody groups (e.g. 64 out of 128),
```
```suggestion
	// Semi-supernode mode custodies at least half of all custody groups (e.g. 64 out of 128),
```

prestonvanloon commented on beacon-chain/sync/custody.go:
```diff
@@ -103,12 +103,35 @@ func (s *Service) custodyGroupCount(context.Context) (uint64, error) {
+		"effectiveCustodyRequirement":  effectiveCustodyRequirement,
+		"custodyRequirement":           cfg.CustodyRequirement,
+		"validatorsCustodyRequirement": validatorsCustodyRequirement,
+	}).Warn("Custody requirements exceed semi-supernode mode; using higher custody count to meet requirements.")
```
Is this worth a WARN level log? If running with many validators, this log message would be firing often?

james-prysm replied:
This is a good point, I think it'll fire off 1once a minute after reivew I should make sure it only triggers once

james-prysm replied:
after trying a few things it seems overly complex for this, just turned it to debug. let me know if it's preferred to just remove it.

james-prysm replied:
i ended up removing it https://github.com/OffchainLabs/prysm/pull/16029/commits/a6a67b5427135a8392fe6619c6b4539fe529b705

james-prysm committed:
Update beacon-chain/sync/custody.go

Co-authored-by: Preston Van Loon <pvanloon@offchainlabs.com>

james-prysm committed:
Update cmd/beacon-chain/flags/config.go

Co-authored-by: Preston Van Loon <pvanloon@offchainlabs.com>

james-prysm committed:
switched it to debug

nalepae commented on cmd/beacon-chain/flags/base.go:
```diff
@@ -339,6 +339,11 @@ var (
+	// SemiSupernode enables subscription to half of all data subnets and custody of 64 data columns.
+	SemiSupernode = &cli.BoolFlag{
+		Name:  "semi-supernode",
+		Usage: "Enable subscription to half of all data subnets and custody 64 data columns (minimum required for reconstruction). Blobs are reconstructed as needed from the custodied 64 columns. If validators require higher custody (based on total stake), the node automatically scales up to meet those requirements with a warning. This is effectively one-way: once enabled, the node keeps storing and se ...
```
Is this possible to break this string in multiple, shorter lines?

james-prysm replied:
https://github.com/OffchainLabs/prysm/pull/16029/commits/7646fc961234f4ae38c7b47f88c5d02154618683 addressed here

nalepae commented on beacon-chain/sync/custody.go:
```diff
@@ -103,12 +103,35 @@ func (s *Service) custodyGroupCount(context.Context) (uint64, error) {
+
+	// Semi-supernode mode custodies at least half of all custody groups (e.g. 64 out of 128),
+	// which is the minimum required for blob reconstruction.
+	semiSupernodeTarget := cfg.NumberOfCustodyGroups / 2
```
Can you define and use a function equivalent tne
```go
func MinimumColumnCountToReconstruct() uint64
```

(defined in the peerDAS package)

but for subnets?

The purpose of the semi super node is to custody enough data column sidecars to reconstruct the missing ones.
Using such a function will make the node future-proof if one day we change the count of data columns subnets, the count of group, the count of data column sidecars per data column subnets, or the count of data column sidecars.

With the current configuration (both `mainnet` and `minimal`), the subnet count = the group count = the data columns count = 128, so `cfg.NumberOfCustodyGroups / 2` works, but is should also work for other values of the configuration.

james-prysm replied:
hope I addressed it correctly here https://github.com/OffchainLabs/prysm/pull/16029/commits/3ee87712105f33c344c8ff8413d26baee10564f3

nalepae commented on beacon-chain/sync/custody.go:
```diff
@@ -103,12 +103,35 @@ func (s *Service) custodyGroupCount(context.Context) (uint64, error) {
+		"effectiveCustodyRequirement":  effectiveCustodyRequirement,
+		"custodyRequirement":           cfg.CustodyRequirement,
+		"validatorsCustodyRequirement": validatorsCustodyRequirement,
+	}).Debug("Custody requirements exceed semi-supernode mode; using higher custody count to meet requirements.")
```
Do we need to print that every minute?

james-prysm replied:
probably not i guess I should remove this

james-prysm replied:
removed here https://github.com/OffchainLabs/prysm/pull/16029/commits/a6a67b5427135a8392fe6619c6b4539fe529b705

nalepae commented on beacon-chain/sync/subscriber.go:
```diff
@@ -692,26 +692,28 @@ func (s *Service) dataColumnSubnetIndices(primitives.Slot) map[uint64]bool {
 	}
 
+	// Get custody group count - this is the source of truth and already reflects:
+	// - Supernode mode: 128 groups
```
Nit: Be careful of the hardcoded value (even in comments)
128 groups is a spec parameters: `NUMBER_OF_CUSTODY_GROUPS`. This value is currently 128, but it could change.

james-prysm replied:
https://github.com/OffchainLabs/prysm/pull/16029/commits/9394750c1aa045ecc0965f396459348faa8e6a5e fixed here

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -471,6 +471,7 @@ func (s *Service) removeStartupState() {
 // It returns the (potentially updated) custody group count and the earliest available slot.
 func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot, uint64, error) {
 	isSubscribedToAllDataSubnets := flags.Get().SubscribeAllDataSubnets
```
While we are here, could we rename `SubscribeAllDataSubnets` to `Supernode`? (And `isSubscribedToAllDataSubnets` to `isSemiSuperNode`)?

james-prysm replied:
https://github.com/OffchainLabs/prysm/pull/16029/commits/700894ef9d18b2f66f0f21052d8b7fb319e8512b addressed here

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -490,10 +491,42 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
+	// Check if the node was previously in semi-supernode mode.
+	wasSemiSupernode, err := s.cfg.BeaconDB.UpdateSemiSupernode(s.ctx, isSemiSupernode)
+	if err != nil {
+		log.WithError(err).Error("Could not update semi-supernode status")
```
I think you re-used the same patterm here (aka, logging if error) because I did it previously for `s.cfg.BeaconDB.UpdateSubscribedToAllDataSubnets`.
However, I cannot remember a good reason (for here and for `s.cfg.BeaconDB.UpdateSubscribedToAllDataSubnets`) why we should log something and continue instead of returning the error. Do you see one?

james-prysm replied:
maybe not didn't think that much of it just followed the same pattern, if it continues we don't want to stop the system from continuing i think?

james-prysm replied:
i ended up removing it for this pr https://github.com/OffchainLabs/prysm/pull/16029/commits/027f6f0b4850b6bf920d5353b52d2189bea31781

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -490,10 +491,42 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
 	custodyGroupCount := custodyRequirement
-	if isSubscribedToAllDataSubnets {
+	if isSubscribedToAllDataSubnets || wasSubscribedToAllDataSubnets {
+		// Full supernode: custody all 128 groups
```
Full supernode = supernode?

james-prysm replied:
yess will remove the full word

james-prysm replied:
fixed in https://github.com/OffchainLabs/prysm/pull/16029/commits/9394750c1aa045ecc0965f396459348faa8e6a5e

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -490,10 +491,42 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
 		custodyGroupCount = cfg.NumberOfCustodyGroups
+	} else if isSemiSupernode || wasSemiSupernode {
+		// Semi-supernode: custody at least 64 groups (half), but use validator requirements if higher
+		semiSupernodeCustody := cfg.NumberOfCustodyGroups / 2
```
Same comment about using the dedicated function to compute how many groups we should custody to enable reconstruction.

james-prysm replied:
https://github.com/OffchainLabs/prysm/pull/16029/commits/3ee87712105f33c344c8ff8413d26baee10564f3 used the function recommended form the other comment

nalepae commented:
When running with both `--semi-supernode` and `--supernode` flags, we get:

```
[2025-11-24 12:03:51.47]  WARN flags: Subscribing to all data subnets
[2025-11-24 12:03:51.47]  WARN flags: Operating in semi-supernode mode (retain 50% of data columns)
```

The node does not effectively operate in semi-supernode mode, but in supernode mode.

Commit:
Add mutual exclusivity validation for --supernode and --semi-supernode flags

james-prysm commented:
> When running with both `--semi-supernode` and `--supernode` flags, we get:
> 
> ```
> [2025-11-24 12:03:51.47]  WARN flags: Subscribing to all data subnets
> [2025-11-24 12:03:51.47]  WARN flags: Operating in semi-supernode mode (retain 50% of data columns)
> ```
> 
> The node does not effectively operate in semi-supernode mode, but in supernode mode.

https://github.com/OffchainLabs/prysm/pull/16029/commits/91a4661561787693dd6548816df0031b0c83273d solved here

james-prysm committed:
remove hardcoded numbers based on feedback

james-prysm committed:
removing not very useful log based on feedback

james-prysm committed:
renaming variables for SubscribeAllDataSubnets to Supernode

james-prysm committed:
breaking up description for flag into multiple lines

james-prysm committed:
reverting db add for semi-supernode while keeping supernodes behavior as the same not to break users, prestons' feedback

james-prysm commented on beacon-chain/blockchain/service.go:
```diff
@@ -470,30 +470,36 @@ func (s *Service) removeStartupState() {
-	}
-
-	// Warn the user if the node was previously subscribed to all data subnets and is not any more.
-	if wasSubscribedToAllDataSubnets && !isSubscribedToAllDataSubnets {
```
moved this down a bit

james-prysm committed:
add recommended MinimumCustodyGroupCountToReconstruct() function

nalepae commented on cmd/beacon-chain/flags/config.go:
```diff
@@ -52,12 +54,20 @@ func ConfigureGlobalFlags(ctx *cli.Context) error {
+
+	// Ensure mutual exclusivity between supernode and semi-supernode modes
+	if supernodeSet && semiSupernodeSet {
+		return fmt.Errorf("cannot set both --%s and --%s flags; choose one mode", SubscribeAllDataSubnets.Name, SemiSupernode.Name)
```
Should it be `Error` (the node continues to operate) or `Fatal` (the node stops operating).
IMO it should be `Fatal`.

Also, in the CLI help, IMO it should be noticed that both options cannot be used together.

james-prysm replied:
throwing an error will stop the node from starting all together

james-prysm replied:
https://github.com/OffchainLabs/prysm/pull/16029/commits/70bfe54012c8aa0cc6079011b5144b264bfb8ead update to description

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -470,30 +471,36 @@ func (s *Service) removeStartupState() {
+
+	// Compute the target custody group count based on current flag configuration.
+	targetCustodyGroupCount := custodyRequirement
+	if isSupernode || wasSupernode {
```
`|| wasSupernode` is not needed, right?

james-prysm replied:
if they take the flag off it should still act like a supernode

james-prysm replied:
I think wasSupernode is sitll required right i'll remove the log and use the simplified max

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -470,30 +471,36 @@ func (s *Service) removeStartupState() {
+		return 0, 0, errors.Wrap(err, "update subscribed to all data subnets")
+	}
+
+	// Compute the target custody group count based on current flag configuration.
```
IMO this part could be simplified by something like:

```go
if isSupernode {
    targetCustodyGroupCount = cfg.NumberOfCustodyGroups
}

if isSemiSupernode {
    targetCustodyGroupCount = max(custodyRequirement, semiSupernodeCustody)
}
```

(No need to add a warning message.)

james-prysm replied:
updated https://github.com/OffchainLabs/prysm/pull/16029/commits/fa9079c63e021beae25338bb1744b075bd30191e but kept the wasSupernode

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -510,12 +517,38 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
 	}
 
-	return earliestAvailableSlot, custodyGroupCount, nil
+	if !wasSupernode && isSupernode && actualCustodyGroupCount < cfg.NumberOfCustodyGroups {
```
Why not only:
```go
if isSupernode {
    ...
}
```
?

This way, this log is printed at each boot when the `--supernode` flag is used.

james-prysm replied:
updated https://github.com/OffchainLabs/prysm/pull/16029/commits/fa9079c63e021beae25338bb1744b075bd30191e

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -510,12 +517,38 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
+			"currentTarget":      targetCustodyGroupCount,
+			"numberOfSubnets":    actualCustodyGroupCount,
+			"dataColumnsPerSlot": actualCustodyGroupCount,
+		}).Warn("Retaining higher custody count from prior configuration. " +
```
Not sure it should be a warning.
This log will be displayed at every start even without using the `--supernode` flag or the `--semi-supernode` flag, as long as there is some active validators connected.
==> This is perfectly normal, so warning may be too scary.

james-prysm replied:
changed to info https://github.com/OffchainLabs/prysm/pull/16029/commits/fa9079c63e021beae25338bb1744b075bd30191e

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -510,12 +517,38 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
+	if wasSupernode && !isSupernode {
+		log.WithField("flag", flags.SubscribeAllDataSubnets.Name).Warn(
+			"Because the flag was previously used, the node will still subscribe to all data subnets.",
+		)
```
```suggestion
		log.Warningf("Because the `--%f` flag was previously used, the node will still subscribe to all data subnets.", flags.SubscribeAllDataSubnets.Name)
```

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -510,12 +517,38 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
+		log.WithFields(logrus.Fields{
+			"current":            actualCustodyGroupCount,
+			"target":             cfg.NumberOfCustodyGroups,
+			"numberOfSubnets":    cfg.NumberOfCustodyGroups,
```
key: number of subnets. value: number of custody groups

james-prysm replied:
removed https://github.com/OffchainLabs/prysm/pull/16029/commits/fa9079c63e021beae25338bb1744b075bd30191e

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -510,12 +517,38 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
+			"current":            actualCustodyGroupCount,
+			"target":             cfg.NumberOfCustodyGroups,
+			"numberOfSubnets":    cfg.NumberOfCustodyGroups,
+			"dataColumnsPerSlot": cfg.NumberOfCustodyGroups,
```
key: data columns per slot. value: number of custody groups.

james-prysm replied:
removed https://github.com/OffchainLabs/prysm/pull/16029/commits/fa9079c63e021beae25338bb1744b075bd30191e

nalepae commented on beacon-chain/core/peerdas/reconstruction.go:
```diff
@@ -29,6 +29,27 @@ func MinimumColumnCountToReconstruct() uint64 {
+	minimumColumnCount := MinimumColumnCountToReconstruct()
+
+	// Calculate how many columns each custody group represents
+	columnsPerGroup := cfg.NumberOfColumns / cfg.NumberOfCustodyGroups
```
Will panic if `cfg.NumberOfCustodyGroups == 0`

james-prysm replied:
added here https://github.com/OffchainLabs/prysm/pull/16029/commits/31c622987f294e4c5dd9133bb69eead99adda24d

nalepae commented on beacon-chain/core/peerdas/reconstruction.go:
```diff
@@ -29,6 +29,27 @@ func MinimumColumnCountToReconstruct() uint64 {
+
+	// Special case: if there are more groups than columns (columnsPerGroup = 0),
+	// we need at least as many groups as we need columns
+	if columnsPerGroup == 0 {
```
This configuration is not realistic. It should error in such a case.

james-prysm replied:
if there are more groups than columns? we wouldn't have something like that in a devnet?

james-prysm replied:
i added change here https://github.com/OffchainLabs/prysm/pull/16029/commits/31c622987f294e4c5dd9133bb69eead99adda24d

nalepae commented on beacon-chain/sync/custody.go:
```diff
@@ -103,12 +103,27 @@ func (s *Service) custodyGroupCount(context.Context) (uint64, error) {
+		return semiSupernodeTarget, nil
+	}
+
+	return effectiveCustodyRequirement, nil
```
```suggestion
	return max(effectiveCustodyRequirement, semiSupernodeTarget), nil
```

Cf. Preston's modernisation PR

james-prysm committed:
Update custody.go

Co-authored-by: Manu NALEPA <enalepa@offchainlabs.com>

james-prysm committed:
Update service.go

Co-authored-by: Manu NALEPA <enalepa@offchainlabs.com>

Commit:
updating feedback for updateCustodyInfoInDB

Commit:
update flag descriptions to indicate mutual exclusivity

Commit:
add error handling to MinimumCustodyGroupCountToReconstruct for invalid config values

nalepae committed:
`updateCustodyInfoInDB`: Removing `|| wasSupernode`.

Reason:

If the node was previously a supernode, then the  `actualCustodyGroupCount` in

`earliestAvailableSlot, actualCustodyGroupCount, err := s.cfg.BeaconDB.UpdateCustodyInfo(s.ctx, slot, targetCustodyGroupCount)`

will return at least the value pushed by `--semi-supernode` (or higher if more than 2048 ETH were managed.)

==> Checking `|| wasSupernode` is useless.

nalepae committed:
`updateCustodyInfoInDB`: Improve warning log.

nalepae committed:
Improve `semiSupernodeSet` log message.

nalepae committed:
Improve `supernodeSet` log message.

nalepae committed:
`cfg.SubscribeAllsubnets` ==> `cfg.Supernode`.

nalepae committed:
`SubscribeAllDataSubets` ==> `Supernode` and refactor help.

nalepae committed:
Change Warning log to Info.

This way, if the node was used as a super node and now the `--semi-supernode` flag is used, we have:

```
[2025-11-26 15:43:51.24]  INFO flags: Operating in semi-supernode mode (custody just enough data to serve the blobs and blob sidecars beacon API)
[2025-11-26 15:43:54.01]  WARN blockchain: Because the `--supernode` flag was previously used, the node will continue to act as a super node.
```

Before, it was two WARNINGs contradicting each others.

nalepae committed:
Remove a log which was misleading, because it would be displayed regardless the `--supernode` or `--semi-supernode` flag usage, as long as some validators are connected.

Specifically, the same log would have been printed even after a restart with a fresh database.

nalepae commented:
Tested with, in this order:
1. no flag
2. `--semi-supernode`
3. `--semi-supernode` again
4. no flag again
5. `--semi-supernode` again
6. `--supernode`
7. no flag again
8. `--semi-supernode` again
9. `--supernode` again
